Project/rag_engine.py

```
@@ -0,0 +1,636 @@
+            nodes.append(node)
+
+    except (sqlite3.OperationalError, sqlite3.DatabaseError) as e:
+        # 2. Fallback to in-memory cosine similarity (Robustness)
+        cursor.execute("SELECT * FROM knowledge_nodes")
+        all_meta = cursor.fetchall()
+
+        all_vecs = {}
+        try:
+            cursor.execute("SELECT node_id, embedding FROM vec_knowledge")
+            all_vecs = {r['node_id']: r['embedding'] for r in cursor.fetchall()}
+        except:
```

Ruff (0.15.6)

[error] Do not use bare `except`

(E722)